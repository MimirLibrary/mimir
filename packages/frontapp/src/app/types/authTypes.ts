import { IUserPayload } from '../store/slices/userSlice';

export type TAuthResponseData = Omit<IUserPayload, 'locations'> & {
  location_id: number;
  location: string;
};
