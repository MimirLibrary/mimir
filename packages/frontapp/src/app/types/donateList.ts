interface IPerson {
  __typename?: 'Person' | undefined;
  id: string;
  username: string;
  avatar: string;
}

interface IStatus {
  __typename?: 'Status' | undefined;
  id: string;
  status: string;
  person: IPerson;
}

export interface IMaterialDonate {
  __typename?: 'Material' | undefined;
  id: string;
  title: string;
  statuses: Array<IStatus | null>;
}
