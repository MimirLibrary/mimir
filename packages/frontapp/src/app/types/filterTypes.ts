export type ParamsType = {
  [key: string]: string[];
};
