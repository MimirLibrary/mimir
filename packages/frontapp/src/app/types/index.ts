export * from './materialsTypes';
export * from './authTypes';
export * from './metadata';
export * from './overdueType';
export * from './notificationList';
