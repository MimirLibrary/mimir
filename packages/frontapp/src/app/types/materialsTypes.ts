import {
  BlockedUsers,
  Location,
  Maybe,
  Message,
  Notification,
  Scalars,
  Status,
} from '@mimir/apollo-client';

export interface IStatus {
  __typename?: 'Status';
  id: string;
  created_at: any;
  status: string;
  person?: IPerson;
}

export interface IMaterial {
  __typename?: 'Material';
  title: string;
  created_at: any;
  picture?: string | null;
  author: string;
  category: string;
  id: string;
  statuses: Array<IStatus | null>;
}
interface IPerson {
  __typename?: 'Person' | undefined;
  id: string;
  username: string;
}

export interface IMaterialsState {
  searchMaterials: Array<IMaterial | null> | null | undefined;
}
export interface IReader {
  __typename?: 'Person';
  avatar?: string | null;
  id: string;
  statuses?: Array<IStatus | null | undefined> | null;
  username: string;
}
export interface IReadersState {
  searchReaders: Array<IReader | null> | null | undefined;
}
