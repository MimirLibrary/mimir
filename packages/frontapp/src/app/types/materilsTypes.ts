export interface IStatus {
  __typename?: 'Status';
  id: string;
  created_at: any;
  status: string;
}

export interface IMaterialState {
  __typename?: 'Material';
  title: string;
  created_at: any;
  picture?: string | null;
  author: string;
  category: string;
  id: string;
  statuses: Array<IStatus | null>;
}

export interface IStateMaterials {
  searchMaterials: Array<IMaterialState | null> | null | undefined;
}
