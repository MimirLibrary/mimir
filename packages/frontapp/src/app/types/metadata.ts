interface IAuthor {
  id: number;
  name: string;
}

interface IPublisher {
  id: number;
  name: string;
}

interface IMeta {
  ageRestriction: string;
  coverType: string;
  dimensions: string;
  manufacturer: string;
  mass: string;
  numberOfPages: string;
  price: string;
  series: string;
  sku: string;
}

interface IMaterialMetaData {
  authors: Array<IAuthor>;
  cover: string;
  description: string;
  title: string;
  yearPublishedAt: number;
  meta: IMeta;
  publisher: IPublisher;
}

export interface IMetaOfMaterial {
  idType: string;
  value: string;
  material: IMaterialMetaData;
}
