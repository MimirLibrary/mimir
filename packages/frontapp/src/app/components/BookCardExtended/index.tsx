import React, { FC, useMemo } from 'react';
import styled from '@emotion/styled';
import { colors, dimensions } from '@mimir/ui-kit';
import src from '../../../assets/MOC-data/BookImage.png';
import { getCurrentStatus } from '../../models/helperFunctions/getCurrentStatus';
import { IMaterial } from '../../types';
import { useNavigate } from 'react-router-dom';
import { OpenLink } from '../ManagerInfoCard';
import { RoutesTypes } from '../../../utils/routes';
import { StatusTypes } from '@mimir/global-types';

const Wrapper = styled.div`
  background: ${colors.bg_secondary};
  border-radius: ${dimensions.xs_1};
  transition: box-shadow 0.3s;
  padding: ${dimensions.xl_2};
  display: flex;
  max-width: 300px;
  width: 100%;
  cursor: pointer;
  :hover {
    box-shadow: 0px 6px 14px -6px rgba(24, 39, 75, 0.08),
      0px 10px 32px -4px rgba(24, 39, 75, 0.08);
  }
`;

const WrapperImg = styled.div`
  img {
    width: 72px;
    height: 115px;
  }
`;

const WrapperDescription = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  margin-left: ${dimensions.base};
`;

const Title = styled.h4`
  font-weight: 500;
  font-size: ${dimensions.base};
  color: ${colors.main_black};
  line-height: ${dimensions.xl};
  margin-bottom: ${dimensions.base};
`;

const TitleGenre = styled.p`
  font-weight: 300;
  font-size: ${dimensions.sm};
  color: ${colors.description_gray};
  line-height: ${dimensions.base};
  margin-bottom: 0.25rem;
`;

const TitleState = styled.p`
  font-weight: 500;
  font-size: ${dimensions.sm};
  color: ${colors.main_black};
  line-height: ${dimensions.base};
`;

const TitleStatus = styled.span`
  font-weight: 300;
  font-size: ${dimensions.sm};
  color: ${colors.main_black};
  line-height: ${dimensions.base};
  margin-bottom: 0.25rem;
  margin-right: 0.2rem;
`;

const StyledUserName = styled(OpenLink)<{ type: string }>`
  text-decoration: underline;
  font-weight: 300;
  font-size: ${dimensions.sm};
  color: ${(props) =>
    props.type === 'Overdue' ? colors.problem_red : colors.accent_color};
`;

const TitleClaimHistory = styled.p`
  font-weight: 500;
  font-size: ${dimensions.sm};
  color: ${colors.main_black};
  line-height: ${dimensions.base};
  margin-top: ${dimensions.xs_2};
  margin-bottom: 0.25rem;
`;

const HistoryBook = styled.p`
  font-weight: 300;
  font-size: ${dimensions.sm};
  color: ${colors.main_black};
  line-height: ${dimensions.base};
`;

interface IPropsBookCardExtended {
  item: IMaterial | null;
}

const BookCardExtended: FC<IPropsBookCardExtended> = ({ item }) => {
  const countOfHistoryClaimed = item?.statuses.filter(
    (elem) => elem?.status === StatusTypes.BUSY
  ).length;
  const navigate = useNavigate();
  const currenStatusElement = item?.statuses[item?.statuses.length - 1];
  const currentStatus = useMemo(
    () => getCurrentStatus(currenStatusElement),
    [currenStatusElement]
  );

  const handleItemRedirect = () => {
    navigate(`/item/${item?.id}`);
  };

  return (
    item && (
      <Wrapper onClick={handleItemRedirect} data-testid="book-card-extended">
        <WrapperImg>
          <img src={item?.picture || src} alt="book-img" />
        </WrapperImg>
        <WrapperDescription>
          <Title>{item?.title}</Title>
          <TitleGenre>{item?.category}</TitleGenre>
          <TitleState>State:</TitleState>
          <TitleStatus>
            {typeof currentStatus === 'string' ? (
              currentStatus
            ) : (
              <>
                <TitleStatus>
                  {currentStatus.type === 'Overdue'
                    ? 'Overdue by'
                    : 'Claimed by'}
                </TitleStatus>
                <StyledUserName
                  type={currentStatus.type}
                  to={`${RoutesTypes.READERS}/${currentStatus.person_id}`}
                >
                  {currentStatus.body}
                </StyledUserName>
              </>
            )}
          </TitleStatus>
          <TitleClaimHistory>Claim history:</TitleClaimHistory>
          <HistoryBook>
            was claimed {countOfHistoryClaimed || 0} times
          </HistoryBook>
        </WrapperDescription>
      </Wrapper>
    )
  );
};

export default BookCardExtended;
