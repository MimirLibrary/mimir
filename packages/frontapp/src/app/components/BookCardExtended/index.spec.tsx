import { render } from '../../../helpers/customRender';
import BookCardExtended from './index';
import { IMaterial } from '../../types';
import { StatusTypes } from '@mimir/global-types';
import { screen } from '@testing-library/react';

describe('render BookCardExtended', () => {
  it('correctly render with status is free', () => {
    const mockItem: IMaterial = {
      author: 'Ivan',
      category: 'Horror',
      title: 'Dracula',
      created_at: '2022-09-21T08:29:24.192Z',
      id: '1',
      picture: '',
      statuses: [
        {
          id: '1',
          created_at: '2022-09-21T08:29:24.192Z',
          status: StatusTypes.FREE,
          person: {
            id: '1',
            username: 'Dmitry',
          },
        },
      ],
    };
    render(<BookCardExtended item={mockItem} />);

    expect(screen.getByTestId('book-card-extended')).toBeInTheDocument();
    expect(screen.getByTestId('book-card-extended')).toMatchSnapshot();
    expect(screen.getByText('Dracula')).toBeInTheDocument();
    expect(screen.getByText('Horror')).toBeInTheDocument();
    expect(screen.getByText('on the shelf')).toBeInTheDocument();
    expect(screen.getByText('was claimed 0 times')).toBeInTheDocument();
  });

  it('correctly render with status is Busy', () => {
    const mockItem: IMaterial = {
      author: 'Ivan',
      category: 'Horror',
      title: 'Dracula',
      created_at: new Date().toDateString(),
      id: '1',
      picture: '',
      statuses: [
        {
          id: '1',
          created_at: new Date().toDateString(),
          status: StatusTypes.BUSY,
          person: {
            id: '1',
            username: 'Dmitry',
          },
        },
      ],
    };
    render(<BookCardExtended item={mockItem} />);

    expect(screen.getByTestId('book-card-extended')).toBeInTheDocument();
    expect(screen.getByText('Dracula')).toBeInTheDocument();
    expect(screen.getByText('Horror')).toBeInTheDocument();
    expect(screen.getByText('Claimed by')).toBeInTheDocument();
    expect(screen.getByText('Dmitry')).toBeInTheDocument();
    expect(screen.getByText('was claimed 1 times')).toBeInTheDocument();
  });

  it('correctly render result with item is null', () => {
    render(<BookCardExtended item={null} />);
    expect(screen.queryByTestId('book-card-extended')).toBeNull();
  });
});
