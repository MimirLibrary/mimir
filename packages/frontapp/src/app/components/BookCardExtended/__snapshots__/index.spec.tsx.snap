// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render BookCardExtended correctly render with status is free 1`] = `
.emotion-0 {
  background: #FFFFFF;
  border-radius: 0.625rem;
  -webkit-transition: box-shadow 0.3s;
  transition: box-shadow 0.3s;
  padding: 1.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-width: 300px;
  width: 100%;
  cursor: pointer;
}

.emotion-0:hover {
  box-shadow: 0px 6px 14px -6px rgba(24, 39, 75, 0.08),0px 10px 32px -4px rgba(24, 39, 75, 0.08);
}

.emotion-1 img {
  width: 72px;
  height: 115px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  margin-left: 1rem;
}

.emotion-3 {
  font-weight: 500;
  font-size: 1rem;
  color: #333333;
  line-height: 1.25rem;
  margin-bottom: 1rem;
}

.emotion-4 {
  font-weight: 300;
  font-size: 0.875rem;
  color: #4F4F4F;
  line-height: 1rem;
  margin-bottom: 0.25rem;
}

.emotion-5 {
  font-weight: 500;
  font-size: 0.875rem;
  color: #333333;
  line-height: 1rem;
}

.emotion-6 {
  font-weight: 300;
  font-size: 0.875rem;
  color: #333333;
  line-height: 1rem;
  margin-bottom: 0.25rem;
  margin-right: 0.2rem;
}

.emotion-7 {
  font-weight: 500;
  font-size: 0.875rem;
  color: #333333;
  line-height: 1rem;
  margin-top: 0.5rem;
  margin-bottom: 0.25rem;
}

.emotion-8 {
  font-weight: 300;
  font-size: 0.875rem;
  color: #333333;
  line-height: 1rem;
}

<div
  class="emotion-0"
  data-testid="book-card-extended"
>
  <div
    class="emotion-1"
  >
    <img
      alt="book-img"
      src="BookImage.png"
    />
  </div>
  <div
    class="emotion-2"
  >
    <h4
      class="emotion-3"
    >
      Dracula
    </h4>
    <p
      class="emotion-4"
    >
      Horror
    </p>
    <p
      class="emotion-5"
    >
      State:
    </p>
    <span
      class="emotion-6"
    >
      on the shelf
    </span>
    <p
      class="emotion-7"
    >
      Claim history:
    </p>
    <p
      class="emotion-8"
    >
      was claimed 
      0
       times
    </p>
  </div>
</div>
`;
